etna/models/nn/deepstate/deepstate.py

```diff
@@ -0,0 +1,317 @@
+from typing import Any
+from typing import Dict
+from typing import Iterator
+from typing import Optional
+
+import pandas as pd
+import torch
+import torch.nn as nn
+from torch import Tensor
+from typing_extensions import TypedDict
+
+from etna.models.base import DeepBaseModel
+from etna.models.base import DeepBaseNet
+from etna.models.nn.deepstate import LDS
+from etna.models.nn.deepstate import CompositeSSM
+
+
+class DeepStateBatch(TypedDict):
+    """Batch specification for DeepStateModel."""
+
+    encoder_real: Tensor  # (batch_size, seq_length, input_size)
+    decoder_real: Tensor  # (batch_size, horizon, input_size)
+    datetime_index: Tensor  # (batch_size, num_models , seq_length + horizon)
+    encoder_target: Tensor  # (batch_size, seq_length, 1)
+
+
+class DeepStateNet(DeepBaseNet):
+    """DeepState network."""
+
+    def __init__(
+        self,
+        ssm: CompositeSSM,
+        input_size: int,
+        num_layers: int,
+        n_samples: int,
+        lr: float,
+        optimizer_params: Optional[dict],
+    ):
+        """Create instance of DeepStateNet.
+
+        Parameters
+        ----------
+        ssm:
+            State Space Model of the system.
+        input_size:
+            Size of the input feature space: target plus extra features.
+        num_layers:
+            Number of layers in RNN.
+        n_samples:
+            Number of samples to use in predictions generation.
+        lr:
+            Learning rate.
+        optimizer_params:
+            Parameters for optimizer for Adam optimizer (api reference :py:class:`torch.optim.Adam`)
+        """
+        super().__init__()
+        self.ssm = ssm
+        self.input_size = input_size
+        self.num_layers = num_layers
+        self.n_samples = n_samples
+        self.lr = lr
+        self.optimizer_params = {} if optimizer_params is None else optimizer_params
+        self.latent_dim = self.ssm.latent_dim()
+
+        self.RNN = nn.LSTM(
+            num_layers=self.num_layers, hidden_size=self.latent_dim, input_size=self.input_size, batch_first=True
+        )
+        self.projectors = nn.ModuleDict(
```

Comment:
In the paper there is also the third type of transformations for the bounded parameters -- may be we can create the separate issue to add this option + prediction intervals